dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramStageWorkingListObjectBundleHookTest.java

```diff
@@ -0,0 +1,325 @@
+
+    private ProgramStageWorkingList programStageWorkingList;
+
+    @BeforeEach
+    public void setUp()
+    {
+        workingListHook = new ProgramStageWorkingListObjectBundleHook( dataElementService,
+            organisationUnitService, attributeService );
+        programStageWorkingList = new ProgramStageWorkingList();
+    }
+
+    @Test
+    void shouldReturnNoErrorsWhenQueryCriteriaSuppliedIsValid()
+    {
+        EventDataFilter dataFilter = new EventDataFilter();
+        dataFilter.setDataItem( "DataItem" );
+        AttributeValueFilter attributeValueFilter = new AttributeValueFilter();
+        attributeValueFilter.setAttribute( "attribute" );
+        ProgramStageQueryCriteria queryCriteria = ProgramStageQueryCriteria.builder()
+            .enrolledAt( createDatePeriod( DateTime.now().minusDays( 1 ).toDate(),
+                DateTime.now().plusDays( 1 ).toDate(), DatePeriodType.ABSOLUTE ) )
+            .enrollmentOccurredAt( createDatePeriod( DateTime.now().minusDays( 1 ).toDate(),
+                DateTime.now().plusDays( 1 ).toDate(), DatePeriodType.RELATIVE ) )
+            .eventCreatedAt( createDatePeriod( DateTime.now().minusDays( 1 ).toDate(),
+                DateTime.now().plusDays( 1 ).toDate(), DatePeriodType.ABSOLUTE ) )
+            .scheduledAt( createDatePeriod( DateTime.now().minusDays( 1 ).toDate(),
+                DateTime.now().plusDays( 1 ).toDate(), DatePeriodType.RELATIVE ) )
+            .assignedUsers( Collections.singleton( "User" ) )
+            .assignedUserMode( AssignedUserSelectionMode.PROVIDED )
+            .orgUnit( "orgUnit" )
+            .ouMode( OrganisationUnitSelectionMode.SELECTED )
+            .dataFilters( List.of( dataFilter ) )
+            .attributeValueFilters( Collections.singletonList( attributeValueFilter ) )
+            .build();
+
+        when( organisationUnitService.getOrganisationUnit( anyString() ) ).thenReturn( new OrganisationUnit() );
+        when( dataElementService.getDataElement( anyString() ) ).thenReturn( new DataElement() );
+        when( attributeService.getTrackedEntityAttribute( anyString() ) ).thenReturn( new TrackedEntityAttribute() );
+        programStageWorkingList.setProgramStageQueryCriteria( queryCriteria );
+
+        List<ErrorReport> errorReports = workingListHook.validate( programStageWorkingList, bundle );
+
+        assertEquals( 0, errorReports.size() );
+    }
+
+    @Test
+    void shouldFailWhenDatePeriodAbsoluteAndNoStartNorEndDateSpecifiedForEnrolledAt()
+    {
+        ProgramStageQueryCriteria queryCriteria = ProgramStageQueryCriteria.builder()
+            .enrolledAt( createDatePeriod( null, null, DatePeriodType.ABSOLUTE ) )
+            .build();
+        programStageWorkingList.setProgramStageQueryCriteria( queryCriteria );
+
+        List<ErrorReport> errorReports = workingListHook.validate( programStageWorkingList, bundle );
+
+        assertEquals( 1, errorReports.size() );
```

Comment:
I don't think we should use this pattern 😅 As explained this leads to error messages that leave out important context. You will get `expected 2 to be 1` (or similar). You will not know whether the one you expected is in or not and what other errors we might have gotten. Tracker assertions https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/validator/AssertValidations.java try to provide that context.

I would also try to avoid asserting on the entire error message. If someone just wants to fix a typo it would be great not having to fix tests as well.

If there is not yet a helper for asserting `ErrorReport` we should create one. For example

```java
    @Test
    void test() {
        ErrorReport error1 = new ErrorReport( OrganisationUnit.class, ErrorCode.E1102, "foo" );
        ErrorReport error2 = new ErrorReport( Period.class, ErrorCode.E1101, "bar" );
        List<ErrorReport> errors = List.of(error1, error2);

        assertErrorReport(errors, ErrorCode.E1102, "wrong arg"); // passes with 'foo'
    }

    private void assertErrorReport(List<ErrorReport> actual, ErrorCode expectedCode, String expectedMessage) {
        assertFalse(actual.isEmpty(), expectedCode + " not found as there are no errors");
        assertTrue( actual.stream().anyMatch( err -> expectedCode == err.getErrorCode() &&
                        err.getMessage().contains( expectedMessage )),
                () -> String.format( "error report with code %s and and (partial) message '%s' not found in %s", expectedCode, expectedMessage, actual ));
    }
```

gives an error like

```
org.opentest4j.AssertionFailedError: error report with code E1102 and and (partial) message 'wrong arg' not found in [ErrorReport{message=Organisation unit not found or not accessible: `foo`, errorCode=E1102, mainKlass=class org.hisp.dhis.organisationunit.OrganisationUnit, errorKlass=null, value=null}, ErrorReport{message=Period is invalid: `bar`, errorCode=E1101, mainKlass=class org.hisp.dhis.period.Period, errorKlass=null, value=null}] ==> 
Expected :true
Actual   :false
```

If you absolutely want to make sure there is only x amount of errors you can add that assertion into the helper wrapping the assertions that belong together in an `assertAll`. This way we get all the context we need 😄